### ISSUE TITLE
GDScript clearing function state deletes local variables

### ISSUE THREAD

**Issue by afiefh** [godotengine/godot#44863]
**Godot version:**
Godot_v3.2.3-stable_x11.64 

**OS/device including version:**
Kubuntu 20.10 laptop with NVidia graphics (issue is not graphics related.)

**Issue description:**
Clearing a variable that contains a function state currently executing clears the local variable within the executed function as well.

**Steps to reproduce:**
* Assign function state to Node's variable.
* In the coroutine have some local variables.
* Resume function state.
* From within the resumed function state, set Node's variable to null

Expected: Function state is referenced by the execution and therefore its local variables are not deleted.
Actual behavior: local variables within function state are cleared when null is assigned.

**Minimal reproduction project:**
[testGDScriptExecution.zip](https://github.com/godotengine/godot/files/5759393/testGDScriptExecution.zip)

**Comment by Calinou** [comment]
Note that `yield()` and function states have been removed in favor of `await` in the `master` branch. This means this bug may not be fixed before 4.0 is released.

**Comment by KoBeWi** [comment]
This is 3.2-specific issue. Function state is not even available in GDScript 2.0.

**Comment by arrowinaknee** [comment]
I can reproduce this issue with the code below:
```GDScript
var function_state

func test():
	var a = 5
	yield()
	print(a)
	function_state = null
	print(a)

func _ready():
	function_state = test()
	if function_state != null and function_state.is_valid():
		function_state.resume()
```
The output of this code is:
```
5
Null
```
But if I add `function_state.reference()` after `yield()` the function state does not get deleted and the local variables remain in place. This workaround also applies to the example provided by @afiefh.
So the fix seems to be pretty simple to me - reference the function state when calling/resuming the function and unreference it when yielding or returning. Or am I missing something?

**Comment by afiefh** [comment]
Yes, I believe the engine should take a reference to the GDFunctionState when it is executing it to avoid having it deleted while executing.

I wasn't aware that function_states have a `reference()` method as that's not documented in the [Godot API](https://docs.godotengine.org/en/2.1/classes/class_gdfunctionstate.html). I might be able to work around the issue using this, but there's quite a bit of state management in my game so it might get messy to know which state needs to be referenced at each point.

I hope GDScript 2.0 keeps a way to write generators/coroutines. The little documentation I was able to find seems to indicate that the new `await` implementation is more geared towards signal based resuming as opposed to today's caller based, but I might just not have seen the relevant part of the documentation/implementation.

**Comment by arrowinaknee** [comment]
`reference()` comes from `Reference` class that `GDFunctionState` extends. But manually referencing a function state in the function that it represents and also keeping the state in a local variable to `unreference()` it before exiting the function is too cumbersome, this definetly should be done by the engine itself.
